File: matterjs-server-main/python_client/tests/helpers/matter_test_client.py
"""Extended MatterClient for integration testing.

Ports the functionality of the JS MatterTestClient.ts, providing event
capture and waiting utilities on top of the real Python MatterClient.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from matter_server.client import MatterClient
from matter_server.common.errors import MatterError

if TYPE_CHECKING:
    from collections.abc import Callable

    import aiohttp

    from matter_server.common.models import EventType


@dataclass
class WsEvent:
    """A captured WebSocket event for test assertions."""

    event: str
    data: Any


@dataclass
class EventWaiter:
    """A pending event waiter with optional matcher predicate."""

    event_type: str
    matcher: Callable[[Any], bool] | None
    future: asyncio.Future[WsEvent]


class MatterTestClient(MatterClient):
    """MatterClient extended with testing utilities.

    Provides event capture, event waiting, and convenience methods
    for integration test scenarios.
    """

    def __init__(self, url: str, session: aiohttp.ClientSession):
        super().__init__(url, session)
        self._events: list[WsEvent] = []
        self._waiters: list[EventWaiter] = []
        # Subscribe to all events (no filter = wildcard) to capture them
        self._unsubscribe = self.subscribe_events(callback=self._on_event)

    def _on_event(self, event: EventType, data: Any) -> None:
        """Capture all events for test assertions and resolve pending waiters."""
        ws_event = WsEvent(event=event.value, data=data)
        self._events.append(ws_event)
        # Check if any waiters match this event
        for waiter in list(self._waiters):
            if waiter.event_type == ws_event.event:
                if waiter.matcher is None or waiter.matcher(ws_event.data):
                    self._waiters.remove(waiter)
                    if not waiter.future.done():
                        waiter.future.set_result(ws_event)

    def clear_events(self) -> None:
        """Clear all collected events."""
        self._events.clear()

    def get_events(self) -> list[WsEvent]:
        """Return a copy of collected events."""
        return list(self._events)

    async def wait_for_event(
        self,
        event_type: str,
        matcher: Callable[[Any], bool] | None = None,
        timeout: float = 10.0,
    ) -> WsEvent:
        """Wait for a specific event, checking already-captured events first.

        Args:
            event_type: The event type string to wait for (e.g. "node_added").
            matcher: Optional predicate to filter event data.
            timeout: Maximum time to wait in seconds.

        Returns:
            The matching WsEvent.

        Raises:
            TimeoutError: If no matching event arrives within the timeout.
        """
        # Check already-captured events first
        for event in self._events:
            if event.event == event_type:
                if matcher is None or matcher(event.data):
                    return event

        # Wait for a new matching event
        loop = asyncio.get_running_loop()
        future: asyncio.Future[WsEvent] = loop.create_future()
        waiter = EventWaiter(event_type=event_type, matcher=matcher, future=future)
        self._waiters.append(waiter)

        try:
            return await asyncio.wait_for(future, timeout=timeout)
        except TimeoutError:
            if waiter in self._waiters:
                self._waiters.remove(waiter)
            raise TimeoutError(f"Timeout waiting for event: {event_type}") from None

    async def connect_and_get_server_info(self):
        """Connect to the server and return the ServerInfoMessage.

        Returns:
            The ServerInfoMessage received on connection.
        """
        await self.connect()
        return self.server_info

    async def start_listening_and_get_nodes(self):
        """Start listening and return the initial node list.

        Starts the listening loop as a background task and waits for
        the initial node dump to be received.

        Returns:
            A tuple of (nodes_list, listen_task) where listen_task is the
            asyncio.Task running the listen loop. The caller is responsible
            for cancelling the task when done.
        """
        init_ready = asyncio.Event()
        listen_task = asyncio.create_task(self.start_listening(init_ready))
        await asyncio.wait_for(init_ready.wait(), timeout=30)
        nodes = self.get_nodes()
        return nodes, listen_task

    async def fetch_server_info(self):
        """Fetch server info via the server_info API command.

        Returns:
            The raw result from the server_info command.
        """
        return await self.send_command("server_info")

    async def close(self) -> None:
        """Disconnect and cleanup.

        Convenience method that calls disconnect() and waits briefly
        for the connection to fully close.
        """
        await self.disconnect()
        await asyncio.sleep(0.1)

    async def send_command_expect_error(
        self, command: str, args: dict | None = None
    ) -> dict:
        """Send a command and expect it to raise an error.

        Args:
            command: The API command string to send.
            args: Optional keyword arguments for the command.

        Returns:
            A dict with 'error_code' and 'details' from the exception.

        Raises:
            AssertionError: If the command succeeds instead of raising an error.
        """
        try:
            await self.send_command(command, **(args or {}))
            raise AssertionError(f"Expected error for command {command}")
        except MatterError as e:
            return {"error_code": e.error_code, "details": str(e)}
        except Exception as e:
            return {"error_code": getattr(e, "error_code", -1), "details": str(e)}

File: matterjs-server-main/packages/matter-server/test/helpers/MatterTestClient.ts
/**
 * @license
 * Copyright 2025-2026 Open Home Foundation
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * Test client extending MatterClient with test-specific functionality.
 * Provides event waiting, event collection, and additional test utilities.
 */

import {
    CommissionableNodeData,
    EventMessage,
    MatterClient,
    MatterNode,
    type MatterNodeData,
    type MatterNodeEvent,
    ServerInfoMessage,
    WebSocketLike,
} from "@matter-server/ws-client";
import type { ServerErrorCode } from "@matter-server/ws-controller";
import WebSocket from "ws";

/** Error with error code from server */
export interface ServerErrorResponse {
    error_code: ServerErrorCode;
    details: string;
}

/** WebSocket event for testing */
export interface WsEvent {
    event: string;
    data: unknown;
}

interface EventWaiter {
    eventType: string;
    matcher?: (data: unknown) => boolean;
    resolve: (event: WsEvent) => void;
    reject: (error: Error) => void;
}

/**
 * Node.js WebSocket factory for MatterClient.
 */
function createNodeWebSocket(url: string): WebSocketLike {
    return new WebSocket(url) as unknown as WebSocketLike;
}

/**
 * Extended MatterClient for testing purposes.
 * Adds event waiting, event collection, and additional test utilities.
 */
export class MatterTestClient extends MatterClient {
    private events: WsEvent[] = [];
    private eventWaiters: EventWaiter[] = [];

    constructor(url: string) {
        super(url, createNodeWebSocket);
    }

    /**
     * Connect to the server and return server info.
     * Unlike the base class, this returns the ServerInfoMessage directly.
     */
    async connectAndGetServerInfo(): Promise<ServerInfoMessage> {
        await this.connect();
        return this.serverInfo;
    }

    /**
     * Discover commissionable nodes (alias for discoverCommissionableNodes).
     */
    async discover(): Promise<CommissionableNodeData[]> {
        return this.discoverCommissionableNodes();
    }

    /**
     * Get diagnostics from the server.
     */
    async getDiagnostics(): Promise<{ info: ServerInfoMessage; nodes: MatterNodeData[]; events: MatterNodeEvent[] }> {
        return this.sendCommand("diagnostics", 0, {});
    }

    /**
     * Commission a device on the network with a setup PIN code.
     */
    async commissionOnNetwork(
        setupPinCode: number,
        filterType?: number,
        filter?: number,
        ipAddr?: string,
    ): Promise<MatterNode> {
        const data = await this.sendCommand("commission_on_network", 0, {
            setup_pin_code: setupPinCode,
            filter_type: filterType,
            filter,
            ip_addr: ipAddr,
        });
        return new MatterNode(data);
    }

    /**
     * Set Thread operational dataset (alias with different name).
     */
    async setThreadDataset(dataset: string): Promise<void> {
        await this.setThreadOperationalDataset(dataset);
    }

    /**
     * Get node IP addresses (camelCase alias).
     */
    async getNodeIpAddresses(nodeId: number | bigint, preferCache = false, scoped = false): Promise<string[]> {
        return this.getNodeIPAddresses(nodeId, preferCache, scoped);
    }

    /**
     * Clear collected events.
     */
    clearEvents(): void {
        this.events = [];
    }

    /**
     * Get all collected events.
     */
    getEvents(): WsEvent[] {
        return [...this.events];
    }

    /**
     * Get the cached server info (returns null if not connected).
     */
    getServerInfo(): ServerInfoMessage | null {
        try {
            return this.serverInfo;
        } catch {
            return null;
        }
    }

    /**
     * Wait for a specific event with optional matcher and timeout.
     */
    async waitForEvent(eventType: string, matcher?: (data: unknown) => boolean, timeoutMs = 10_000): Promise<WsEvent> {
        // Check existing events first
        const existing = this.events.find(e => e.event === eventType && (!matcher || matcher(e.data)));
        if (existing) {
            return existing;
        }

        // Wait for new event
        return new Promise((resolve, reject) => {
            const timeout = setTimeout(() => {
                const idx = this.eventWaiters.findIndex(w => w.resolve === resolve);
                if (idx >= 0) {
                    this.eventWaiters.splice(idx, 1);
                }
                reject(new Error(`Timeout waiting for event: ${eventType}`));
            }, timeoutMs);

            this.eventWaiters.push({
                eventType,
                matcher,
                resolve: event => {
                    clearTimeout(timeout);
                    resolve(event);
                },
                reject: error => {
                    clearTimeout(timeout);
                    reject(error);
                },
            });
        });
    }

    /**
     * Close the connection.
     */
    async close(): Promise<void> {
        this.disconnect();
        // Give a moment for cleanup
        await new Promise(resolve => setTimeout(resolve, 100));
    }

    /**
     * Start listening and return the initial nodes array.
     * This is a convenience method for testing that returns nodes.
     */
    async startListeningAndGetNodes(): Promise<MatterNode[]> {
        await super.startListening();
        return Object.values(this.nodes);
    }

    /**
     * Send a command and expect it to fail with a specific error code.
     * Uses a separate raw WebSocket connection to capture the full error response.
     */
    async sendCommandExpectError(
        command: string,
        args: Record<string, unknown>,
        timeoutMs = 10_000,
    ): Promise<ServerErrorResponse> {
        // Create a temporary raw WebSocket connection for this test
        const url = this.connection.ws_server_url;

        return new Promise((resolve, reject) => {
            const ws = new WebSocket(url);
            const messageId = `error_test_${Math.random().toString(36).substring(7)}`;
            let serverInfoReceived = false;

            const timeout = setTimeout(() => {
                ws.close();
                reject(new Error(`Timeout waiting for error response to ${command}`));
            }, timeoutMs);

            ws.onopen = () => {
                // Wait for server info first, then send command
            };

            ws.onmessage = event => {
                const data = JSON.parse(event.data as string, (_key, value) => {
                    // Handle BigInt serialization
                    if (typeof value === "string" && /^\d{16,}$/.test(value)) {
                        return BigInt(value);
                    }
                    return value;
                });

                if (!serverInfoReceived) {
                    // First message is server info
                    serverInfoReceived = true;
                    // Now send our test command
                    ws.send(
                        JSON.stringify({
                            message_id: messageId,
                            command,
                            args,
                        }),
                    );
                    return;
                }

                // Check if this is our response
                if (data.message_id === messageId) {
                    clearTimeout(timeout);
                    ws.close();

                    if ("error_code" in data) {
                        resolve({
                            error_code: data.error_code,
                            details: data.details ?? "",
                        });
                    } else {
                        reject(new Error(`Expected error response but got success for ${command}`));
                    }
                }
            };

            ws.onerror = () => {
                clearTimeout(timeout);
                reject(new Error("WebSocket error"));
            };
        });
    }

    /**
     * Hook into event handling to collect events for testing.
     * This is called by the parent class for each incoming event.
     */
    protected override onRawEvent(event: EventMessage): void {
        const wsEvent: WsEvent = {
            event: event.event,
            data: event.data,
        };
        this.events.push(wsEvent);

        // Check if any waiters match this event
        for (let i = this.eventWaiters.length - 1; i >= 0; i--) {
            const waiter = this.eventWaiters[i];
            if (waiter.eventType === wsEvent.event && (!waiter.matcher || waiter.matcher(wsEvent.data))) {
                this.eventWaiters.splice(i, 1);
                waiter.resolve(wsEvent);
            }
        }
    }
}
